File: clients/python/examples/tournament.py
"""Tournament ranking: rate teams from head-to-head games.

Run after `maturin develop`:  python examples/tournament.py
"""

import propagon


def main() -> None:
    games = propagon.GamesDataset()
    # winner, loser (each game is one observation)
    for winner, loser in [
        ("Sharks", "Bears"),
        ("Sharks", "Wolves"),
        ("Bears", "Wolves"),
        ("Wolves", "Bears"),
        ("Sharks", "Bears"),
    ]:
        games.push_pair(winner, loser)

    # Incremental rating systems
    for algo in (propagon.Elo(k=24.0), propagon.Glicko2()):
        model = algo.fit(games)
        print(f"\n{type(algo).__name__}:")
        for name, score in model.sorted_scores():
            print(f"  {name:8s} {score:8.2f}")

    # Glicko-2 also exposes rating deviation and volatility per team.
    print("\nGlicko-2 detail (rating ± 2·RD):")
    model = propagon.Glicko2().fit(games)
    for name, state in model.players():
        lo, hi = state.bounds()
        print(f"  {name:8s} {state.r:7.1f}  [{lo:.0f}, {hi:.0f}]  sigma={state.sigma:.4f}")


if __name__ == "__main__":
    main()

File: clients/wasm/examples/tournament.ts
// Tournament ratings from head-to-head games: Elo and Glicko-2.
//
// Build first (npm run build), then run with Node 24+ type stripping:
//   node examples/tournament.ts
import { datasets, games } from "../index.js";

const d = new datasets.GamesDataset();
for (const [winner, loser] of [
  ["sharks", "bears"],
  ["sharks", "wolves"],
  ["bears", "wolves"],
  ["wolves", "bears"],
  ["sharks", "bears"],
]) {
  d.pushPair(winner, loser, 1);
}

console.log("Elo:");
for (const [name, score] of games.fitElo({ k: 24 }, d).sortedScores()) {
  console.log(`  ${name.padEnd(8)} ${score.toFixed(1)}`);
}

console.log("\nGlicko-2 (rating ± 2·RD):");
const g = games.fitGlicko2({}, d);
for (const [name, state] of g.players()) {
  console.log(
    `  ${name.padEnd(8)} ${state.r.toFixed(0)} ± ${(2 * state.rd).toFixed(0)}`,
  );
}
